src/System.Net.Security/src/System/Net/Security/SslStreamPal.Windows.cs

```diff
@@ -81,6 +82,22 @@ public static SecurityStatusPal InitializeSecurityContext(ref SafeFreeCredential
             return SecurityStatusAdapterPal.GetSecurityStatusPalFromNativeInt(errorCode);
         }
 
+        public unsafe static string GetServerIdentity(byte[] clientHello)
+        {
+            fixed (byte* clientHelloBytes = clientHello)
+            {
+                byte* serverIdentityBytes;
+                int serverIdentitySize;
+                int ret = Interop.SChannel.SslGetServerIdentity(clientHelloBytes, clientHello.Length, out serverIdentityBytes, out serverIdentitySize, 0);
+                if (ret != 0 || serverIdentitySize == 0)
+                {
+                    return null;
+                }
+
+                return Encoding.UTF8.GetString(serverIdentityBytes, serverIdentitySize);
```

Comment:
Pls verify RFC for TLS/SSL that it is UTF8 encoding.

Reply:
still pending

Reply:
@davidsh
RFC 4366 defines UTF-8 (should only be used with host names which are defined to be UTF-8)
RFC 6066 (App. A) overrides it to ASCII encoding for both host name and SNI. We should stick to whatever we do for host name

Reply:
slightly confused on the code -not seeing any other encoding than utf-8 in our code base and targetName for client side seems to be delegated to SspiCli - we should check somehow what that does

Reply:
I believe that it is correct at one stage utf8 was okay. But as ASCII is the one true way now for DNS names (with internationalization being supported via App A) and this isn't an existing feature you shouldn't provide backwards compat. Although I do wonder if you should just surface a Span<byte> for checking anyway

Reply:
More fun with punycode? Are you going to un-escape that before invoking the callback or is the caller going to have to deal with that?

Reply:
thanks @Drawaes, @Tratcher - I'm not planning on un-escaping anything - either bytes or straight forward conversion.
I'm going to start with implementing other platforms and see what APIs they expose - if all of them expose bytes I'll consider using Span instead but for the client we expose string which might make it confusing - on the other hand this is fairly low level API so it might be ok.

Reply:
I would tend to agree that if you are writing a server that needs its own TLS and you are supporting SNI then you will mostly likely understand the issues and this should be low level (or escaped)

Reply:
IDN is not widely understood. We need to A) test with common clients like browsers, and B) document the results.

Reply:
From a usability perspective we'd rather surface the decoded version to developers so they don't have to worry about it.

Reply:
I agree that the string returned from the API should be "decoded" meaning it should undo any Punnycode if it was transmitted that way on the wire.

Reply:
@Tratcher: I have mixed feelings about this. From dev perspective it is always easier to do less. From perf perspective it is better to not create a copy of the string (the name is already allocated inside client hello) and comparing can be done with byte comparison (assuming names are in the same encoding which should be the case). From TLS spec standpoint it probably deoesn't matter as it is simply a byte array and length and encoding is an implementation detail (probably that's why they are being vague about what should implementation do about it)

Reply:
You're not avoiding any work in the end, you're only punting it to the caller.

Reply:
You kinda are. If you have a server name you would eacape it 1x on the server. Then when a name comes in from the client you do a byte compare. You would likely do the initial escape once and the compare for n number of connections. If you have to depunifiy© the client to compare to a string everytime then you are doing n units of work.

Reply:
Fair, but it requires the caller to be IDN aware.